mtls-northbound/README.md

```diff
@@ -0,0 +1,603 @@
+
+## Setup instructions
+
+1. Clone the apigee-samples repo, and switch to the mtls-northbound directory
+
+```bash
+git clone https://github.com/GoogleCloudPlatform/apigee-samples.git
+cd apigee-samples/mtls-northbound
+```
+
+2. Ensure you have an active GCP account selected in the Cloud shell
+
+```sh
+gcloud auth login
+```
+
+3. Edit the `env.sh` and configure the ENV vars specific to your installation.
+
+* `PROJECT` the project where your Apigee organization is located.
+* `APIGEE_HOST` the externally reachable hostname of the Apigee environment group that contains APIGEE_ENV.
+* `APIGEE_ENV` the Apigee environment where the demo resources should be created.
+* `POOL` the name of the Private CA pool (e.g. partners-pool).
+* `ROOT` the name of the Root CA in the pool (e.g. partner1-root-ca).
+* `TRUST_CONFIG` the name of the Trust Configuration for the Root CA (e.g. partner1-trust-config).
+* `CERT_NAME` the name of the client certificate used in API calls (e.g. partner-1-client-1).
+
+Other environment variables set below.
+* `TARGET_PROXY` the Target HTTPS Proxy in your GLB configuration (set below).
+* `BACKEND_SERVICE` the Backend Service in your GLB configuration (set below).
+
+Now source the `env.sh` file
+
+```bash
+source ./env.sh
+```
+
+## Create CA pool and root CAs
+We'll use the use case of multiple partners communicating with our APIs.
+
+### CA Pool
+Create a pool for each of the partner CAs.
+```
+gcloud privateca pools create ${POOL} --location=us-east1
+```
+
+### Root CAs
+Create a Root CA.
+```
+gcloud privateca roots create ${ROOT} \
+    --pool=${POOL} \
+    --subject="CN=partner1-ca, O=Partner 1" \
+    --location=us-east1 \
+    --auto-enable
+```
+
+## Create Trust Configuration
+The trust configuration can be used by multiple Server TLS Policies (e.g. lenient, strict).
+
+First, get the private CA root certificate.
+```
+gcloud privateca roots describe ${ROOT} \
+   --pool=${POOL} \
+   --location=us-east1 \
+   --format='value(pemCaCertificates)' > ${ROOT}.cert
+```
+Then format to remove newlines and create the trust config yaml file. 
+```
+export ROOT_PEM=$(cat ${ROOT}.cert | sed 's/^[ ]*//g' | tr '\n' $ | sed 's/\$/\\n/g')
+
+cat << EOF > ${TRUST_CONFIG}.yaml
+name: ${TRUST_CONFIG}
+trustStores:
+- trustAnchors:
+   - pemCertificate: "${ROOT_PEM?}"
+EOF
+```
+Finally create the trust config.
+```
+gcloud beta certificate-manager trust-configs import ${TRUST_CONFIG} --source=${TRUST_CONFIG}.yaml
+```
+
+## Create Server TLS Policies
+Create policy representations for "lenient" and "strict", 
+then you can easily switch between them by updating the Target HTTPS Proxy.
+
+### Lenient
+Create the lenient policy yaml file.
+```
+cat << EOF > ${ROOT}-lenient.yaml
+name: ${ROOT}-lenient
+mtlsPolicy:
+  clientValidationMode: ALLOW_INVALID_OR_MISSING_CLIENT_CERT
+  clientValidationTrustConfig: projects/${PROJECT}/locations/global/trustConfigs/${TRUST_CONFIG}
+EOF
+```
+Create the lenient policy.
```

Comment:
`networksecurity.googleapis.com` API should be enabled first

Reply:
OK